src/gui/transferlistwidget.cpp

```diff
@@ -607,6 +607,65 @@ void TransferListWidget::askNewCategoryForSelection()
     } while(invalid);
 }
 
+void TransferListWidget::askAddTagsForSelection()
+{
+    const QStringList tags = askTagsForSelection(tr("Add Tags"));
+    foreach (const QString &tag, tags)
+        addSelectionTag(tag);
+}
+
+void TransferListWidget::askRemoveTagsForSelection()
+{
+    const QStringList tags = askTagsForSelection(tr("Remove Tags"));
+    foreach (const QString &tag, tags) {
+        removeSelectionTag(tag);
+    }
+}
+
+void TransferListWidget::confirmRemoveAllTagsForSelection()
+{
+    QMessageBox::StandardButton response = QMessageBox::question(
+        this, tr("Remove All Tags"), tr("Remove all tags from selected torrents?"),
+        QMessageBox::Yes | QMessageBox::No);
+    if (response == QMessageBox::Yes)
+        clearSelectionTags();
+}
+
+QStringList TransferListWidget::askTagsForSelection(const QString &dialogTitle)
+{
+    QStringList tags;
+    bool invalid = true;
+    while (invalid) {
+        bool ok = false;
+        invalid = false;
+        const QString tagsInput = AutoExpandableDialog::getText(
+            this, dialogTitle, tr("Comma-separated tags:"), QLineEdit::Normal, "", &ok).trimmed();
+        if (!ok || tagsInput.isEmpty())
+            return QStringList();
+        tags = tagsInput.split(',', QString::SkipEmptyParts);
+        for (QString &tag : tags) {
+            tag = tag.trimmed();
+            if (!BitTorrent::Session::isValidTag(tag)) {
+                QMessageBox::warning(this, tr("Invalid tag")
+                                     , tr("Tag name: '%1' is invalid").arg(tag));
+                invalid = true;
+            }
+        }
+    }
+    return tags;
+}
+
+void TransferListWidget::applyToSelectedTorrents(const std::function<void(BitTorrent::TorrentHandle *const)> &fn)
+{
+    foreach (const QModelIndex &index, selectionModel()->selectedRows()) {
+        BitTorrent::TorrentHandle *const torrent = listModel->torrentHandle(mapToSource(index));
+        if (torrent)
```

Comment:
Here we are dealing with a known valid indices, isn't it? Therefore, we assume that each of them corresponds to a valid torrent unless some errors in code. So we should use `Q_ASSERT(torrent);` instead of `if-else`.